File: app/src/androidTest/java/openfoodfacts/github/scrachx/openfood/repositories/ProductRepositoryTest.java
package openfoodfacts.github.scrachx.openfood.repositories;

import android.util.Log;
import androidx.test.filters.SmallTest;
import androidx.test.runner.AndroidJUnit4;
import openfoodfacts.github.scrachx.openfood.models.Allergen;
import openfoodfacts.github.scrachx.openfood.models.AllergenName;
import openfoodfacts.github.scrachx.openfood.models.DaoSession;
import openfoodfacts.github.scrachx.openfood.views.OFFApplication;
import org.greenrobot.greendao.database.Database;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.junit.Assert.*;

/**
 * Created by Lobster on 05.03.18.
 */
@SmallTest
@RunWith(AndroidJUnit4.class)
public class ProductRepositoryTest {
    private static final String TEST_ALLERGEN_TAG = "en:lupin";
    private static final String TEST_LANGUAGE_CODE = "es";
    private static final String TEST_ALLERGEN_NAME = "Altramuces";
    private static IProductRepository productRepository;

    @BeforeClass
    public static void cleanAllergens() {
        clearDatabase();
        productRepository = ProductRepository.getInstance();
        productRepository.saveAllergens(createAllergens());
    }

    private static void clearDatabase() {
        DaoSession daoSession = OFFApplication.getInstance().getDaoSession();
        Database db = daoSession.getDatabase();
        db.beginTransaction();
        try {
            daoSession.getAllergenDao().deleteAll();
            db.setTransactionSuccessful();
        } catch (Exception e) {
            Log.e("openfoodfacts.github.scrachx.openfood.repositories.ProductRepositoryTest", "error in transaction", e);
        } finally {
            db.endTransaction();
        }
    }

    @AfterClass
    public static void close() {
        clearDatabase();
    }

    @Test
    public void testGetAllergens() {
        List<Allergen> allergens = productRepository.getAllergens(false).blockingGet();

        assertNotNull(allergens);
        assertEquals(2, allergens.size());

        Allergen allergen = allergens.get(0);
        assertEquals(TEST_ALLERGEN_TAG, allergen.getTag());

        List<AllergenName> allergenNames = allergen.getNames();
        assertEquals(3, allergenNames.size());

        AllergenName allergenName = allergenNames.get(0);
        assertEquals(allergenName.getAllergenTag(), allergen.getTag());
        assertEquals(TEST_LANGUAGE_CODE, allergenName.getLanguageCode());
        assertEquals(TEST_ALLERGEN_NAME, allergenName.getName());
    }

    @Test
    public void testGetEnabledAllergens() {
        List<Allergen> allergens = productRepository.getEnabledAllergens();

        assertNotNull(allergens);
        assertEquals(1, allergens.size());
        assertEquals(TEST_ALLERGEN_TAG, allergens.get(0).getTag());
    }

    @Test
    public void testGetAllergensByEnabledAndLanguageCode() {
        List<AllergenName> enabledAllergenNames = productRepository.getAllergensByEnabledAndLanguageCode(true, TEST_LANGUAGE_CODE).blockingGet();
        List<AllergenName> notEnabledAllergenNames = productRepository.getAllergensByEnabledAndLanguageCode(false, TEST_LANGUAGE_CODE).blockingGet();

        assertNotNull(enabledAllergenNames);
        assertNotNull(notEnabledAllergenNames);

        assertEquals(1, enabledAllergenNames.size());
        assertEquals(1, notEnabledAllergenNames.size());

        assertEquals(TEST_ALLERGEN_NAME, enabledAllergenNames.get(0).getName());
        assertEquals("Molluschi", notEnabledAllergenNames.get(0).getName());
    }

    @Test
    public void testGetAllergensByLanguageCode() {
        List<AllergenName> allergenNames = productRepository.getAllergensByLanguageCode(TEST_LANGUAGE_CODE).blockingGet();

        assertNotNull(allergenNames);
        assertEquals(2, allergenNames.size());
    }

    private static List<Allergen> createAllergens() {
        Allergen allergen1 = new Allergen(TEST_ALLERGEN_TAG, new ArrayList<>());
        allergen1.setEnabled(true);
        allergen1.getNames().add(new AllergenName(allergen1.getTag(), TEST_LANGUAGE_CODE, TEST_ALLERGEN_NAME));
        allergen1.getNames().add(new AllergenName(allergen1.getTag(), "bg", "Лупина"));
        allergen1.getNames().add(new AllergenName(allergen1.getTag(), "fr", "Lupin"));

        Allergen allergen2 = new Allergen("en:molluscs", new ArrayList<>());
        allergen2.getNames().add(new AllergenName(allergen2.getTag(), TEST_LANGUAGE_CODE, "Molluschi"));
        allergen2.getNames().add(new AllergenName(allergen2.getTag(), "en", "Mollusques"));

        return Arrays.asList(allergen1, allergen2);
    }
}

File: app/src/androidTest/java/openfoodfacts/github/scrachx/openfood/repositories/ProductRepositoryTest.kt
package openfoodfacts.github.scrachx.openfood.repositories

import android.content.Context
import androidx.test.filters.SmallTest
import com.google.common.truth.Truth.assertThat
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.android.testing.HiltAndroidRule
import dagger.hilt.android.testing.HiltAndroidTest
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.rx2.await
import kotlinx.coroutines.test.runBlockingTest
import openfoodfacts.github.scrachx.openfood.models.DaoSession
import openfoodfacts.github.scrachx.openfood.models.entities.allergen.Allergen
import openfoodfacts.github.scrachx.openfood.models.entities.allergen.AllergenName
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import javax.inject.Inject

/**
 * Created by Lobster on 05.03.18.
 */
@ExperimentalCoroutinesApi
@SmallTest
@HiltAndroidTest
class ProductRepositoryTest {
    @get:Rule
    var hiltRule = HiltAndroidRule(this)


    @Inject
    lateinit var productRepository: ProductRepository

    @Inject
    @ApplicationContext
    lateinit var instance: Context

    @Inject
    lateinit var daoSession: DaoSession


    @Before
    fun cleanAllergens() {
        hiltRule.inject()
        clearDatabase(daoSession)
        productRepository.saveAllergens(createAllergens())
    }

    @Test
    fun testGetAllergens() {
        val mSettings = instance.getSharedPreferences("prefs", 0)
        val isDownloadActivated = mSettings.getBoolean(Taxonomy.ALLERGEN.downloadActivatePreferencesId, false)
        val allergens = productRepository.reloadAllergensFromServer().blockingGet()

        assertThat(allergens).isNotNull()

        if (!isDownloadActivated) {
            assertThat(allergens).hasSize(0)
        } else {
            assertThat(allergens).hasSize(2)

            val allergen = allergens[0]
            assertThat(allergen.tag).isEqualTo(TEST_ALLERGEN_TAG)

            val allergenNames = allergen.names
            assertThat(allergenNames).hasSize(3)

            val allergenName = allergenNames[0]
            assertThat(allergen.tag).isEqualTo(allergenName.allergenTag)
            assertThat(allergenName.languageCode).isEqualTo(TEST_LANGUAGE_CODE)
            assertThat(allergenName.name).isEqualTo(TEST_ALLERGEN_NAME)
        }
    }

    @Test
    fun testGetEnabledAllergens() = runBlockingTest {
        val allergens = productRepository.getEnabledAllergens()
        assertThat(allergens).isNotNull()
        assertThat(allergens).hasSize(1)
        assertThat(allergens[0].tag).isEqualTo(TEST_ALLERGEN_TAG)
    }

    @Test
    fun testGetAllergensByEnabledAndLanguageCode() {
        val enabledAllergenNames = productRepository.getAllergensByEnabledAndLanguageCode(true, TEST_LANGUAGE_CODE).blockingGet()
        val notEnabledAllergenNames = productRepository.getAllergensByEnabledAndLanguageCode(false, TEST_LANGUAGE_CODE).blockingGet()

        assertThat(enabledAllergenNames).isNotNull()
        assertThat(notEnabledAllergenNames).isNotNull()

        assertThat(enabledAllergenNames).hasSize(1)
        assertThat(notEnabledAllergenNames).hasSize(1)

        assertThat(enabledAllergenNames[0].name).isEqualTo(TEST_ALLERGEN_NAME)
        assertThat(notEnabledAllergenNames[0].name).isEqualTo("Molluschi")
    }

    @Test
    fun testGetAllergensByLanguageCode() = runBlockingTest {
        val allergenNames = productRepository.getAllergensByLanguageCode(TEST_LANGUAGE_CODE).await()
        assertNotNull(allergenNames)
        assertEquals(2, allergenNames.size.toLong())
    }


    @After
    fun close() = clearDatabase(daoSession)

    companion object {
        private const val TEST_ALLERGEN_TAG = "en:lupin"
        private const val TEST_LANGUAGE_CODE = "es"
        private const val TEST_ALLERGEN_NAME = "Altramuces"

        private fun clearDatabase(daoSession: DaoSession) = daoSession.allergenDao.deleteAll()

        private fun createAllergens(): List<Allergen> {
            var tag = TEST_ALLERGEN_TAG
            val allergen1 = Allergen(tag, listOf(
                    AllergenName(tag, TEST_LANGUAGE_CODE, TEST_ALLERGEN_NAME),
                    AllergenName(tag, "bg", "Лупина"),
                    AllergenName(tag, "fr", "Lupin")
            )).apply {
                enabled = true
            }

            tag = "en:molluscs"
            val allergen2 = Allergen(tag, listOf(
                    AllergenName(tag, TEST_LANGUAGE_CODE, "Molluschi"),
                    AllergenName(tag, "en", "Mollusques")
            ))

            return listOf(allergen1, allergen2)
        }
    }
}